test/tools/cmap_spec_runner.ts

```diff
@@ -0,0 +1,336 @@
+ */
+class Thread {
+  _promise: Promise<void>;
+  _error: Error;
+  _killed = false;
+
+  _knownCommands: any;
+
+  // concurrent execution context
+  constructor(operations) {
+    this._promise = new Promise(resolve => {
+      this.start = () => resolve();
+    });
+
+    this._knownCommands = operations;
+  }
+
+  start: () => void;
+
+  queue(op: cmapOperation) {
+    if (this._killed || this._error) {
+      return;
+    }
+
+    this._promise = this._promise.then(() => this._runOperation(op)).catch(e => (this._error = e));
+  }
+
+  async _runOperation(op: cmapOperation) {
+    const operationFn = this._knownCommands[op.name];
+    if (!operationFn) {
+      throw new Error(`Invalid command ${op.name}`);
+    }
+
+    await operationFn(op);
+    await sleep();
+  }
+
+  async finish() {
+    this._killed = true;
+    await this._promise;
+    if (this._error) {
+      throw this._error;
+    }
+  }
+}
+
+/**
+ * Implements the spec test match function, see:
+ * [CMAP Spec Test README](https://github.com/mongodb/specifications/tree/master/source/connection-monitoring-and-pooling/tests#spec-test-match-function)
+ */
+const compareInputToSpec = (input, expected) => {
+  if (expected === 42) {
+    expect(input).to.be.ok; // not null or undefined
+    return;
+  }
+
+  if (Array.isArray(expected)) {
+    expect(input).to.be.an('array');
+    expected.forEach((expectedValue, index) => {
+      compareInputToSpec(input[index], expectedValue);
+    });
+    return;
+  }
+
+  if (expected && typeof expected === 'object') {
+    for (const [expectedPropName, expectedValue] of Object.entries(expected)) {
+      expect(input).to.have.property(expectedPropName);
+      compareInputToSpec(input[expectedPropName], expectedValue);
+    }
+    return;
+  }
+
+  expect(input).to.equal(expected);
+};
+
+const getTestOpDefinitions = (threadContext: ThreadContext) => ({
+  checkOut: async function (op) {
+    const connection: Connection = await util
+      .promisify(ConnectionPool.prototype.checkOut)
+      .call(threadContext.pool);
+    if (op.label != null) {
+      threadContext.connections.set(op.label, connection);
+    } else {
+      threadContext.orphans.add(connection);
+    }
+  },
+  checkIn: function (op) {
+    const connection = threadContext.connections.get(op.connection);
+    threadContext.connections.delete(op.connection);
+
+    if (!connection) {
+      throw new Error(`Attempted to release non-existient connection ${op.connection}`);
+    }
+
+    return threadContext.pool.checkIn(connection);
+  },
+  clear: function () {
+    return threadContext.pool.clear();
+  },
+  close: function () {
+    return util.promisify(ConnectionPool.prototype.close).call(threadContext.pool);
+  },
+  wait: async function (options) {
+    const ms = options.ms;
+    return sleep(ms);
+  },
+  start: function (options) {
+    const target = options.target;
+    const thread = threadContext.getThread(target);
+    thread.start();
```

Comment:
This is so strange, start is equivalent to calling the resolve defined inside a promise's executor function passed to its constructor. So this isn't async? and shouldn't be awaited? Just trying to follow the logic of what this does... 

Edit: Oh so I think I see it, basically this holds up the start promise, and since our run operations are chained on after this, calling start will unblock them in a sense. Is that right?

Reply:
yup; this logic hasn't changed from the original implementation